forc-test/src/execute.rs

```diff
@@ -23,10 +21,85 @@ use vm::interpreter::{InterpreterParams, MemoryInstance};
 use vm::state::DebugEval;
 use vm::state::ProgramState;
 
+#[derive(Debug, Clone)]
+pub enum CapturedEcal {
+    Write { fd: u64, bytes: Vec<u8> },
+    Unknown { ra: u64, rb: u64, rc: u64, rd: u64 },
+}
+
+impl CapturedEcal {
+    pub fn apply(&self) {
+        match self {
+            CapturedEcal::Write { fd, bytes } => {
+                let s = std::str::from_utf8(bytes.as_slice()).unwrap();
+
+                use std::io::Write;
+                use std::os::fd::FromRawFd;
+
+                let mut f = unsafe { std::fs::File::from_raw_fd(*fd as i32) };
+                write!(&mut f, "{}", s).unwrap();
+
+                // Dont close the fd
+                std::mem::forget(f);
+            }
+            CapturedEcal::Unknown { ra, rb, rc, rd } => {
+                println!("Unknown ecal: {} {} {} {}", ra, rb, rc, rd);
+            }
+        }
+    }
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct EcalState {
```

Comment:
Note: Could use some doc-comments for this - since we seem to be writing to a Vec here (its not a simple ecal impl)

Reply:
Done.